File: s3.go
package main

import (
	"context"
	"net/url"
	"strings"
	"time"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

var minioClient *minio.Client

func initS3Client() error {
	host, secure := parseEndpoint(globalConfig.S3Endpoint)

	var err error
	minioClient, err = minio.New(host, &minio.Options{
		Creds:  credentials.NewStaticV4(globalConfig.S3AccessKeyID, globalConfig.S3SecretAccessKey, ""),
		Secure: secure,
	})
	if err != nil {
		return err
	}
	return nil
}

func parseEndpoint(endpoint string) (string, bool) {
	if endpoint == "" {
		return "s3.amazonaws.com", true
	}
	secure := true
	host := endpoint
	if strings.HasPrefix(endpoint, "https://") {
		host = strings.TrimPrefix(endpoint, "https://")
		secure = true
	} else if strings.HasPrefix(endpoint, "http://") {
		host = strings.TrimPrefix(endpoint, "http://")
		secure = false
	}
	return host, secure
}

// ListDirectory lists files and folders for a given S3 prefix
func ListDirectory(ctx context.Context, prefix string) (DirectoryListing, error) {
	queryPrefix := prefix
	if queryPrefix != "" && !strings.HasSuffix(queryPrefix, "/") {
		queryPrefix = queryPrefix + "/"
	}

	folders := make([]FolderEntry, 0, 32)
	files := make([]FileEntry, 0, 64)

	objectCh := minioClient.ListObjects(ctx, globalConfig.S3Bucket, minio.ListObjectsOptions{
		Prefix:    queryPrefix,
	})

	for obj := range objectCh {
		if obj.Err != nil {
			return DirectoryListing{}, obj.Err
		}

		if strings.HasSuffix(obj.Key, "/") {
			// It is a folder
			name := strings.TrimPrefix(obj.Key, queryPrefix)
			name = strings.TrimSuffix(name, "/")
			if name == "" {
				continue
			}
			folders = append(folders, FolderEntry{Name: name, Path: obj.Key})
		} else {
			// It is a file
			name := strings.TrimPrefix(obj.Key, queryPrefix)
			if name == "" {
				continue
			}
			files = append(files, FileEntry{
				Name:         name,
				Path:         obj.Key,
				Size:         obj.Size,
				LastModified: obj.LastModified.Format(time.RFC3339),
			})
		}
	}

	return DirectoryListing{Folders: folders, Files: files}, nil
}

// GetObjectInfo retrieves metadata (size, contentType, ETag, lastModified) for an object
func GetObjectInfo(ctx context.Context, key string) (FileInfo, error) {
	objInfo, err := minioClient.StatObject(ctx, globalConfig.S3Bucket, key, minio.StatObjectOptions{})
	if err != nil {
		return FileInfo{}, err
	}

	return FileInfo{
		Size:         objInfo.Size,
		ContentType:  objInfo.ContentType,
		LastModified: objInfo.LastModified.Format(time.RFC3339),
		ETag:         objInfo.ETag,
	}, nil
}

// GetPresignedUrl generates a temporary presigned URL for downloading
func GetPresignedUrl(ctx context.Context, key string, expires time.Duration) (string, error) {
	reqParams := make(url.Values)
	u, err := minioClient.PresignedGetObject(ctx, globalConfig.S3Bucket, key, expires, reqParams)
	if err != nil {
		return "", err
	}
	return u.String(), nil
}

// SearchBucket performs a recursive, bucket-wide search matching query
func SearchBucket(ctx context.Context, query string) (SearchResults, error) {
	searchCtx, cancel := context.WithCancel(ctx)
	defer cancel()

	lowerQ := strings.ToLower(query)
	files := make([]FileEntry, 0, 128)
	folderSet := make(map[string]string) // path -> name
	maxKeys := 500

	objectCh := minioClient.ListObjects(searchCtx, globalConfig.S3Bucket, minio.ListObjectsOptions{
		Recursive: true,
	})

	for obj := range objectCh {
		if obj.Err != nil {
			return SearchResults{}, obj.Err
		}

		key := obj.Key
		segments := strings.Split(key, "/")

		// Check folder segments in path
		for i := 0; i < len(segments)-1; i++ {
			seg := segments[i]
			if seg != "" && strings.Contains(strings.ToLower(seg), lowerQ) {
				folderPath := strings.Join(segments[:i+1], "/") + "/"
				if _, ok := folderSet[folderPath]; !ok {
					folderSet[folderPath] = seg
				}
			}
		}

		// Check filename
		if len(segments) > 0 {
			name := segments[len(segments)-1]
			if name != "" && strings.Contains(strings.ToLower(name), lowerQ) {
				files = append(files, FileEntry{
					Name:         name,
					Path:         key,
					Size:         obj.Size,
					LastModified: obj.LastModified.Format(time.RFC3339),
				})
			}
		}

		if len(files) >= maxKeys {
			break
		}
	}

	folders := make([]FolderEntry, 0, len(folderSet))
	for path, name := range folderSet {
		folders = append(folders, FolderEntry{Name: name, Path: path})
	}

	return SearchResults{Files: files, Folders: folders}, nil
}

File: frontend/src/lib/s3.ts
import { S3Client, ListObjectsV2Command, GetObjectCommand, HeadObjectCommand } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';

// To support both Node/Deno (process.env/Deno.env) and Cloudflare Workers (c.env)
export function getS3Client(env: Record<string, string | undefined>) {
  const region = env.S3_REGION || 'auto';
  const endpoint = env.S3_ENDPOINT;
  const accessKeyId = env.S3_ACCESS_KEY_ID || '';
  const secretAccessKey = env.S3_SECRET_ACCESS_KEY || '';

  return new S3Client({
    region,
    endpoint,
    credentials: {
      accessKeyId,
      secretAccessKey,
    },
    // Force path style for S3 compatible storage (like Minio or R2 if needed)
    forcePathStyle: env.S3_FORCE_PATH_STYLE === 'true', 
  });
}

export async function listDirectory(client: S3Client, bucket: string, prefix: string) {
  // Ensure prefix ends with a slash if it's not empty, to correctly list a directory
  const queryPrefix = prefix && !prefix.endsWith('/') ? prefix + '/' : prefix;

  const command = new ListObjectsV2Command({
    Bucket: bucket,
    Prefix: queryPrefix,
    Delimiter: '/', // Delimiter '/' groups objects into "directories" (CommonPrefixes)
  });

  const response = await client.send(command);

  const folders = (response.CommonPrefixes || []).map((p) => ({
    name: p.Prefix?.substring(queryPrefix.length).replace(/\/$/, '') || '',
    path: p.Prefix || '',
  }));

  const files = (response.Contents || [])
    .filter((obj) => obj.Key !== queryPrefix) // Exclude the directory itself if it exists as an object
    .map((obj) => ({
      name: obj.Key?.substring(queryPrefix.length) || '',
      path: obj.Key || '',
      size: obj.Size || 0,
      lastModified: obj.LastModified?.toISOString() || '',
    }));

  return { folders, files };
}

export async function getObjectInfo(client: S3Client, bucket: string, key: string) {
  const command = new HeadObjectCommand({ Bucket: bucket, Key: key });
  try {
    const response = await client.send(command);
    return {
      size: response.ContentLength || 0,
      contentType: response.ContentType || 'application/octet-stream',
      lastModified: response.LastModified?.toISOString() || '',
      eTag: response.ETag || '',
    };
  } catch (error: any) {
    if (error.name === 'NotFound' || error.$metadata?.httpStatusCode === 404) {
      return null;
    }
    throw error;
  }
}

export async function getObject(client: S3Client, bucket: string, key: string, range?: string) {
  const command = new GetObjectCommand({ 
    Bucket: bucket, 
    Key: key,
    ...(range ? { Range: range } : {})
  });
  const response = await client.send(command);
  return response;
}

/** Search all objects in the bucket, returning matching files and folder segments. */
export async function searchBucket(
  client: S3Client,
  bucket: string,
  query: string,
  maxKeys = 500,
) {
  const lowerQ = query.toLowerCase();
  const files: Array<{ name: string; path: string; size: number; lastModified: string }> = [];
  const folderSet = new Map<string, string>(); // path → name

  let continuationToken: string | undefined;

  do {
    const command = new ListObjectsV2Command({
      Bucket: bucket,
      MaxKeys: 1000,
      ContinuationToken: continuationToken,
    });
    const response = await client.send(command);

    for (const obj of response.Contents || []) {
      const key = obj.Key || '';
      const segments = key.split('/');

      // Check each intermediate folder segment
      for (let i = 0; i < segments.length - 1; i++) {
        const seg = segments[i];
        if (seg && seg.toLowerCase().includes(lowerQ)) {
          const folderPath = segments.slice(0, i + 1).join('/') + '/';
          if (!folderSet.has(folderPath)) folderSet.set(folderPath, seg);
        }
      }

      // Check filename
      const name = segments[segments.length - 1];
      if (name && name.toLowerCase().includes(lowerQ)) {
        files.push({
          name,
          path: key,
          size: obj.Size || 0,
          lastModified: obj.LastModified?.toISOString() || '',
        });
      }

      if (files.length >= maxKeys) break;
    }

    continuationToken = response.IsTruncated ? response.NextContinuationToken : undefined;
  } while (continuationToken && files.length < maxKeys);

  const folders = Array.from(folderSet.entries()).map(([path, name]) => ({ name, path }));
  return { files, folders };
}

/** Generate a presigned GET URL for an object in S3, valid for a specified duration. */
export async function getPresignedUrl(
  client: S3Client,
  bucket: string,
  key: string,
  expiresIn = 3600,
) {
  const command = new GetObjectCommand({ Bucket: bucket, Key: key });
  return getSignedUrl(client, command, { expiresIn });
}


